File: demo/flutter/flutternew/lib/service/ali_stt_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:record/record.dart';
import 'package:web_socket_channel/io.dart';

import '../config/module_key_config.dart';

enum AliSttEventType {
  started,
  partial,
  finalResult,
  stopped,
  error,
}

class AliSttEvent {
  final AliSttEventType type;
  final String? text;
  final Object? error;

  AliSttEvent._(this.type, {this.text, this.error});

  factory AliSttEvent.started() => AliSttEvent._(AliSttEventType.started);

  factory AliSttEvent.partial(String text) =>
      AliSttEvent._(AliSttEventType.partial, text: text);

  factory AliSttEvent.finalResult(String text) =>
      AliSttEvent._(AliSttEventType.finalResult, text: text);

  factory AliSttEvent.stopped() => AliSttEvent._(AliSttEventType.stopped);

  factory AliSttEvent.error(Object error) =>
      AliSttEvent._(AliSttEventType.error, error: error);
}

class AliSttService {
  final AudioRecorder _record = AudioRecorder();
  final StreamController<AliSttEvent> _eventController =
      StreamController<AliSttEvent>.broadcast();

  IOWebSocketChannel? _channel;
  StreamSubscription? _channelSub;
  StreamSubscription<List<int>>? _recordSub;

  bool _isRecording = false;
  bool _sessionActive = false;
  bool _channelClosed = false;
  bool _disposed = false;
  bool _taskStarted = false;
  bool _finalEmitted = false;
  String? _taskId;
  Completer<void>? _taskStartedCompleter;

  AliSttKeyConfig? _sttConfig;
  final List<String> _sentenceResults = <String>[];
  String _latestCombinedText = '';

  Stream<AliSttEvent> get events => _eventController.stream;

  Future<void> start() async {
    if (_isRecording || _disposed) {
      return;
    }
    await _channelSub?.cancel();
    _channelSub = null;
    _closeChannel();

    final hasPermission = await _record.hasPermission();
    if (!hasPermission) {
      _eventController.add(AliSttEvent.error('麦克风权限未授权'));
      return;
    }

    _sttConfig ??= (await ModuleKeyConfigStore.load()).sttAli;
    _resetSession();
    _sessionActive = true;
    _isRecording = true;
    _safeEmit(AliSttEvent.started());

    try {
      await _openAndStartTask();
      await _waitTaskStarted();

      final stream = await _record.startStream(
        RecordConfig(
          encoder: AudioEncoder.pcm16bits,
          sampleRate: _sttConfig!.sampleRate,
          numChannels: 1,
        ),
      );
      _recordSub = stream.listen(
        _sendAudioFrame,
        onError: (Object error) => _safeEmit(AliSttEvent.error(error)),
      );
    } catch (e) {
      _safeEmit(AliSttEvent.error(e));
      await _cleanupAfterStartFailure();
    }
  }

  Future<void> stop() async {
    if (!_isRecording && !_sessionActive) {
      return;
    }
    _isRecording = false;
    _sessionActive = false;
    _safeEmit(AliSttEvent.stopped());

    await _recordSub?.cancel();
    _recordSub = null;
    await _record.stop();

    if (_channel == null || _channelClosed) {
      return;
    }
    if (_taskStarted) {
      _sendFinishTask();
    } else {
      _closeChannel();
    }
  }

  Future<void> dispose() async {
    _disposed = true;
    _sessionActive = false;
    _isRecording = false;
    await _recordSub?.cancel();
    await _record.stop();
    await _channelSub?.cancel();
    _closeChannel();
    if (!_eventController.isClosed) {
      await _eventController.close();
    }
  }

  Future<void> _openAndStartTask() async {
    final config = _sttConfig;
    if (config == null) {
      throw const ModuleKeyConfigException('STT_ALI 配置未初始化，请先调用 start()');
    }

    _taskId = _generateTaskId();
    _taskStartedCompleter = Completer<void>();

    _channel = IOWebSocketChannel.connect(
      Uri.parse(config.hostUrl),
      headers: <String, dynamic>{
        HttpHeaders.authorizationHeader: 'bearer ${config.apiKey}',
      },
    );

    _channelSub = _channel!.stream.listen(
      _handleSocketMessage,
      onError: (Object error) => _safeEmit(AliSttEvent.error(error)),
      onDone: () {
        _channelClosed = true;
      },
    );

    _sendRunTask();
  }

  Future<void> _waitTaskStarted() async {
    final completer = _taskStartedCompleter;
    if (completer == null) {
      return;
    }
    await completer.future.timeout(
      const Duration(seconds: 6),
      onTimeout: () {
        throw TimeoutException('阿里STT任务启动超时');
      },
    );
  }

  void _handleSocketMessage(dynamic message) {
    if (message is! String) {
      return;
    }

    Map<String, dynamic> jsonMap;
    try {
      jsonMap = jsonDecode(message) as Map<String, dynamic>;
    } catch (e) {
      _safeEmit(AliSttEvent.error('解析阿里STT返回失败: $e'));
      return;
    }

    final header = (jsonMap['header'] as Map?)?.cast<String, dynamic>() ??
        const <String, dynamic>{};
    final payload = (jsonMap['payload'] as Map?)?.cast<String, dynamic>() ??
        const <String, dynamic>{};
    final event = (header['event'] ?? '').toString();

    switch (event) {
      case 'task-started':
        _taskStarted = true;
        final completer = _taskStartedCompleter;
        if (completer != null && !completer.isCompleted) {
          completer.complete();
        }
        break;
      case 'result-generated':
        _handleResultGenerated(payload);
        break;
      case 'task-finished':
        _emitFinalIfNeeded();
        _closeChannel();
        break;
      case 'task-failed':
        final Object error = 'task-failed: code=${header['error_code'] ?? '-'} '
            'msg=${header['error_message'] ?? '-'}';
        _safeEmit(AliSttEvent.error(error));
        _closeChannel();
        break;
      default:
        if ((header['error_message'] ?? '').toString().isNotEmpty) {
          _safeEmit(
            AliSttEvent.error(
              '阿里STT错误: ${header['error_message']}',
            ),
          );
        }
    }
  }

  void _handleResultGenerated(Map<String, dynamic> payload) {
    final output = (payload['output'] as Map?)?.cast<String, dynamic>() ??
        const <String, dynamic>{};
    final sentence = (output['sentence'] as Map?)?.cast<String, dynamic>() ??
        const <String, dynamic>{};

    if (sentence['heartbeat'] == true) {
      return;
    }

    final text = (sentence['text'] ?? '').toString();
    if (text.isEmpty) {
      return;
    }

    final sentenceEnd = sentence['sentence_end'] == true;
    if (sentenceEnd) {
      _sentenceResults.add(text);
      _latestCombinedText = _sentenceResults.join(' ').trim();
      _safeEmit(AliSttEvent.partial(_latestCombinedText));
      return;
    }

    if (_sentenceResults.isEmpty) {
      _latestCombinedText = text;
    } else {
      _latestCombinedText = '${_sentenceResults.join(' ')} $text'.trim();
    }
    _safeEmit(AliSttEvent.partial(_latestCombinedText));
  }

  void _emitFinalIfNeeded() {
    if (_finalEmitted) {
      return;
    }
    _finalEmitted = true;
    final text = _latestCombinedText.trim();
    _safeEmit(AliSttEvent.finalResult(text));
  }

  void _sendAudioFrame(List<int> data) {
    if (!_isRecording || !_sessionActive || !_taskStarted) {
      return;
    }
    if (_channel == null || _channelClosed || data.isEmpty) {
      return;
    }
    try {
      _channel!.sink.add(Uint8List.fromList(data));
    } catch (_) {
      _channelClosed = true;
    }
  }

  void _sendRunTask() {
    final config = _sttConfig;
    final taskId = _taskId;
    if (config == null || taskId == null || _channel == null || _channelClosed) {
      return;
    }

    final parameters = <String, dynamic>{
      'format': config.format,
      'sample_rate': config.sampleRate,
      'disfluency_removal_enabled': config.disfluencyRemovalEnabled,
    };
    if (config.languageHints.isNotEmpty) {
      parameters['language_hints'] = config.languageHints;
    }

    final runTaskMessage = <String, dynamic>{
      'header': <String, dynamic>{
        'action': 'run-task',
        'task_id': taskId,
        'streaming': 'duplex',
      },
      'payload': <String, dynamic>{
        'task_group': 'audio',
        'task': 'asr',
        'function': 'recognition',
        'model': config.model,
        'parameters': parameters,
        'input': <String, dynamic>{},
      },
    };

    _channel!.sink.add(jsonEncode(runTaskMessage));
  }

  void _sendFinishTask() {
    final taskId = _taskId;
    if (taskId == null || _channel == null || _channelClosed) {
      return;
    }
    final finishTaskMessage = <String, dynamic>{
      'header': <String, dynamic>{
        'action': 'finish-task',
        'task_id': taskId,
        'streaming': 'duplex',
      },
      'payload': <String, dynamic>{
        'input': <String, dynamic>{},
      },
    };

    try {
      _channel!.sink.add(jsonEncode(finishTaskMessage));
    } catch (_) {
      _closeChannel();
    }
  }

  Future<void> _cleanupAfterStartFailure() async {
    _isRecording = false;
    _sessionActive = false;
    await _recordSub?.cancel();
    _recordSub = null;
    await _record.stop();
    await _channelSub?.cancel();
    _closeChannel();
  }

  void _closeChannel() {
    if (!_channelClosed) {
      try {
        _channel?.sink.close();
      } catch (_) {
        // ignore
      }
    }
    _channelClosed = true;
    _channel = null;
  }

  void _resetSession() {
    _channelClosed = false;
    _taskStarted = false;
    _finalEmitted = false;
    _taskId = null;
    _taskStartedCompleter = null;
    _sentenceResults.clear();
    _latestCombinedText = '';
  }

  String _generateTaskId() {
    const String chars = '0123456789abcdef';
    final Random random = Random.secure();
    return List<String>.generate(
      32,
      (_) => chars[random.nextInt(chars.length)],
    ).join();
  }

  void _safeEmit(AliSttEvent event) {
    if (_disposed || _eventController.isClosed) {
      return;
    }
    _eventController.add(event);
  }
}

File: demo/app/app/src/main/java/com/vectordemo/service/voice/AliSttService.kt
package com.vectordemo.service.voice

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.util.Log
import androidx.annotation.RequiresPermission
import androidx.core.content.ContextCompat
import com.vectordemo.config.AliSttKeyConfig
import com.vectordemo.config.ModuleKeyConfigStore
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString.Companion.toByteString
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

enum class AliSttEventType {
    STARTED, PARTIAL, FINAL_RESULT, STOPPED, ERROR
}

data class AliSttEvent(
    val type: AliSttEventType,
    val text: String? = null,
    val error: String? = null
)

class AliSttService(private val context: Context) {
    private companion object {
        const val TAG = "AliSttService"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val client = OkHttpClient.Builder().build()

    private var listener: ((AliSttEvent) -> Unit)? = null
    private var webSocket: WebSocket? = null
    private var wsJob: CompletableDeferred<Unit>? = null
    private var taskId: String? = null
    private var taskStarted = false
    private var finalEmitted = false
    private var running = false
    private var sessionActive = false
    private var taskFailed = false
    private var taskFailedError: String? = null
    private var socketClosed = false
    private var manualClosing = false
    private var audioJob: Job? = null
    private var record: AudioRecord? = null
    private val sentenceResults = mutableListOf<String>()
    private var latestCombinedText: String = ""
    private var audioFramesSent = 0
    private var audioBytesSent = 0
    private var runTaskAtMs = 0L

    fun setListener(listener: ((AliSttEvent) -> Unit)?) {
        this.listener = listener
    }

    fun hasRecordPermission(): Boolean {
        return ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED
    }

    @SuppressLint("MissingPermission")
    suspend fun start() = withContext(Dispatchers.IO) {
        if (running) return@withContext
        if (!hasRecordPermission()) {
            emit(AliSttEvent(AliSttEventType.ERROR, error = "麦克风权限未授权"))
            return@withContext
        }

        val config = ModuleKeyConfigStore.load(context).aliStt
        resetSessionState()
        running = true
        sessionActive = true
        Log.d(TAG, "start: begin new session")
        emit(AliSttEvent(AliSttEventType.STARTED))

        try {
            openAndRunTask(config)
            waitTaskStarted()
            if (!running || !sessionActive || taskFailed || socketClosed || webSocket == null) {
                throw IllegalStateException(taskFailedError ?: "STT会话已中断")
            }
            startRecordingAndSend(config)
            Log.d(TAG, "start: task started, recording loop running")
        } catch (e: Exception) {
            Log.e(TAG, "start failed", e)
            emit(AliSttEvent(AliSttEventType.ERROR, error = e.message ?: "STT启动失败"))
            cleanupOnError()
        }
    }

    suspend fun stop() = withContext(Dispatchers.IO) {
        if (!running) return@withContext
        Log.d(TAG, "stop: taskStarted=$taskStarted frames=$audioFramesSent bytes=$audioBytesSent")
        running = false
        sessionActive = false
        emit(AliSttEvent(AliSttEventType.STOPPED))
        stopRecording()
        if (taskStarted && !socketClosed && webSocket != null) {
            sendFinishTask()
            delay(200)
        } else {
            closeSocket()
        }
    }

    fun release() {
        scope.launch {
            running = false
            sessionActive = false
            stopRecording()
            closeSocket()
            scope.cancel()
        }
    }

    private fun openAndRunTask(config: AliSttKeyConfig) {
        taskId = UUID.randomUUID().toString().replace("-", "")
        wsJob = CompletableDeferred()
        Log.d(TAG, "openAndRunTask: taskId=$taskId url=${config.hostUrl}")
        val request = Request.Builder()
            .url(config.hostUrl)
            .addHeader("Authorization", "bearer ${config.apiKey}")
            .build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "websocket opened, send run-task")
                sendRunTask(config)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleSocketMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                socketClosed = true
                if (manualClosing) {
                    Log.d(TAG, "websocket closed manually")
                    return
                }
                Log.e(TAG, "websocket failure", t)
                wsJob?.completeExceptionally(t)
                if (running || sessionActive) {
                    emit(AliSttEvent(AliSttEventType.ERROR, error = t.message ?: "WebSocket失败"))
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                socketClosed = true
                if (wsJob?.isCompleted == false) {
                    wsJob?.complete(Unit)
                }
            }
        })
    }

    private suspend fun waitTaskStarted() {
        val job = wsJob ?: throw IllegalStateException("STT任务未创建")
        withContext(Dispatchers.IO) {
            kotlinx.coroutines.withTimeout(6000) {
                while (!taskStarted && isActive) {
                    delay(50)
                }
                if (!taskStarted) {
                    throw IllegalStateException(taskFailedError ?: "阿里STT任务启动超时")
                }
            }
        }
        if (job.isCancelled) {
            throw IllegalStateException("阿里STT任务启动失败")
        }
        Log.d(TAG, "waitTaskStarted: success taskId=$taskId")
    }

    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
    private fun startRecordingAndSend(config: AliSttKeyConfig) {
        val channelConfig = AudioFormat.CHANNEL_IN_MONO
        val audioFormat = AudioFormat.ENCODING_PCM_16BIT
        val minBuf = AudioRecord.getMinBufferSize(config.sampleRate, channelConfig, audioFormat)
        val bufferSize = if (minBuf <= 0) 3200 else maxOf(minBuf, 3200)
        val audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            config.sampleRate,
            channelConfig,
            audioFormat,
            bufferSize
        )
        record = audioRecord
        audioRecord.startRecording()

        audioJob = scope.launch {
            val buffer = ByteArray(3200)
            while (isActive && running) {
                val read = audioRecord.read(buffer, 0, buffer.size)
                if (read > 0 && taskStarted && !taskFailed && !socketClosed) {
                    audioFramesSent++
                    audioBytesSent += read
                    webSocket?.send(buffer.copyOf(read).toByteString())
                }
            }
        }
    }

    private fun stopRecording() {
        audioJob?.cancel()
        audioJob = null
        runCatching {
            record?.stop()
        }
        runCatching {
            record?.release()
        }
        record = null
    }

    private fun sendRunTask(config: AliSttKeyConfig) {
        val params = JSONObject()
            .put("format", config.format)
            .put("sample_rate", config.sampleRate)
            .put("disfluency_removal_enabled", config.disfluencyRemovalEnabled)
        if (config.languageHints.isNotEmpty()) {
            params.put("language_hints", JSONArray(config.languageHints))
        }
        val body = JSONObject()
            .put(
                "header",
                JSONObject()
                    .put("action", "run-task")
                    .put("task_id", taskId)
                    .put("streaming", "duplex")
            )
            .put(
                "payload",
                JSONObject()
                    .put("task_group", "audio")
                    .put("task", "asr")
                    .put("function", "recognition")
                    .put("model", config.model)
                    .put("parameters", params)
                    .put("input", JSONObject())
            )
        webSocket?.send(body.toString())
        runTaskAtMs = System.currentTimeMillis()
    }

    private fun sendFinishTask() {
        val body = JSONObject()
            .put(
                "header",
                JSONObject()
                    .put("action", "finish-task")
                    .put("task_id", taskId)
                    .put("streaming", "duplex")
            )
            .put("payload", JSONObject().put("input", JSONObject()))
        webSocket?.send(body.toString())
    }

    private fun handleSocketMessage(message: String) {
        val json = runCatching { JSONObject(message) }.getOrNull() ?: return
        val header = json.optJSONObject("header") ?: JSONObject()
        val payload = json.optJSONObject("payload") ?: JSONObject()
        val event = header.optString("event")
        when (event) {
            "task-started" -> {
                taskStarted = true
                wsJob?.complete(Unit)
                Log.d(TAG, "event task-started: taskId=$taskId, costMs=${System.currentTimeMillis() - runTaskAtMs}")
            }
            "result-generated" -> handleResultGenerated(payload)
            "task-finished" -> {
                Log.d(TAG, "event task-finished: frames=$audioFramesSent bytes=$audioBytesSent")
                emitFinalIfNeeded()
                closeSocket()
            }
            "task-failed" -> {
                val code = header.opt("error_code")
                val msg = header.opt("error_message")
                taskFailed = true
                running = false
                sessionActive = false
                taskFailedError = "task-failed: code=$code msg=$msg"
                if (wsJob?.isCompleted == false) {
                    wsJob?.completeExceptionally(IllegalStateException(taskFailedError))
                }
                Log.e(
                    TAG,
                    "event task-failed: code=$code msg=$msg frames=$audioFramesSent bytes=$audioBytesSent taskStarted=$taskStarted"
                )
                emit(
                    AliSttEvent(
                        AliSttEventType.ERROR,
                        error = "$taskFailedError frames=$audioFramesSent bytes=$audioBytesSent"
                    )
                )
                closeSocket()
            }
            else -> {
                if (event.isNotEmpty()) {
                    Log.d(TAG, "event $event")
                }
            }
        }
    }

    private fun handleResultGenerated(payload: JSONObject) {
        val sentence = payload.optJSONObject("output")
            ?.optJSONObject("sentence")
            ?: return
        if (sentence.optBoolean("heartbeat", false)) {
            return
        }
        val text = sentence.optString("text").orEmpty()
        if (text.isBlank()) {
            return
        }
        val sentenceEnd = sentence.optBoolean("sentence_end", false)
        if (sentenceEnd) {
            sentenceResults.add(text)
            latestCombinedText = sentenceResults.joinToString(" ").trim()
            emit(AliSttEvent(AliSttEventType.PARTIAL, text = latestCombinedText))
            return
        }
        latestCombinedText = if (sentenceResults.isEmpty()) {
            text
        } else {
            "${sentenceResults.joinToString(" ")} $text".trim()
        }
        emit(AliSttEvent(AliSttEventType.PARTIAL, text = latestCombinedText))
    }

    private fun emitFinalIfNeeded() {
        if (finalEmitted) return
        finalEmitted = true
        emit(AliSttEvent(AliSttEventType.FINAL_RESULT, text = latestCombinedText.trim()))
    }

    private fun cleanupOnError() {
        running = false
        sessionActive = false
        stopRecording()
        closeSocket()
    }

    private fun closeSocket() {
        manualClosing = true
        runCatching { webSocket?.close(1000, "close") }
        socketClosed = true
        webSocket = null
    }

    private fun resetSessionState() {
        taskStarted = false
        finalEmitted = false
        taskId = null
        sentenceResults.clear()
        latestCombinedText = ""
        wsJob = null
        taskFailed = false
        taskFailedError = null
        socketClosed = false
        manualClosing = false
        audioFramesSent = 0
        audioBytesSent = 0
        runTaskAtMs = 0L
    }

    private fun emit(event: AliSttEvent) {
        listener?.invoke(event)
    }
}

